### ISSUE TITLE
About custom dataset with TensorRT RuntimeError

### ISSUE THREAD

**Issue by ntut108318099** [haotian-liu/yolact_edge#47]
Sorry to bother you..
I retrained my custom dataset to detect only person class. 

But when I used the model to evaluate,it showed the error about "Expected 3 elements in a list but found 2"...

Did anyone meet the error before ? 

I also used `--use_fp16_tensorrt` command ,and it still get the error.
![image](https://user-images.githubusercontent.com/56862464/106695330-51f09600-6615-11eb-809a-ad48116431f1.png)

I think maybe it's overfitting,so it can't transfer tensorrt properly.

Did it possible ? 

Here's my config.py 
```sh
person_dataset=dataset_base.copy({
    'name': 'person',
    
    'train_info': '/home/jason/dataset/mscoco/train2017_person.json',
    'valid_info': '/home/jason/dataset/mscoco/val2017_person.json',
    'class_names':("person",),
    'label_map': {1:1},
})
```

Thanks for your reading~

**Comment by ntut108318099** [comment]
Sorry to bother you~ 

I retrained the base model (yolact_edge_config) with mscoco 80 classes , and I didn't do any transfer learning.

It just trained with resnet101 model. 

It's also get the error about  "Expected 3 elements in a list but found 2" . 
BTW, I use the latest  original code.  

I think the training part is something wrong...

Did you ever modify the training code? 

Thanks for your supply.

**Comment by haotian-liu** [comment]
This is interesting, thank you for your report. I will be re-training again on the newest code to see if there is anything wrong. Thanks!

**Comment by haotian-liu** [comment]
Btw another note: I have once used the released code to train the fine-tuning models for YolactEdge with video, and the TensorRT conversion works fine. Now I am using the released code to train YolactEdge base model for both COCO and YouTube VIS, and it usually takes 2-3 days. Will report back then. Hopefully we can have this issue resolved ASAP.

**Comment by haotian-liu** [comment]
Hi I have finished training on YouTube VIS dataset with the YolactEdge (without video) model, and the results are good and as expected for both PyTorch version and TensorRT converted version. However, it seems that the one with COCO dataset was not as stable that I saw a irregular loss flipping up at the middle of training. **I wonder if you would be willing to take a look at your Tensorboard logs during training and see if there is such similar loss flipping up.**

I applied a fix in original YOLACT implementation c3a6173 on my dev branch, and will start a training again on COCO tonight. I believe the training code is running fine as expected and there might be some issue related to COCO processing related (not only loading, but also augmentations, GT processing, etc.), and there is similar instability issue for original YOLACT.

Hopefully we can understand the problem ASAP and have it fixed, thanks.

**Comment by ntut108318099** [comment]
Hi, Thanks for your hard work!!

I will try the latest code and make sure the environment correct or not. 
If I have any new progress, I'll let you know immediately.

**Comment by haotian-liu** [comment]
Hi, can you provide me with your training setting, thanks.

(1) number of GPUs used to train the model
(2) batch size per GPU
(3) learning rate
(4) number of iterations trained
(5) number of categories in the dataset
(6) number of samples in the dataset.

**Comment by haotian-liu** [comment]
I somehow figured out that this is related to multiple training instability, and it also exists in original YOLACT https://github.com/dbolya/yolact/issues/586, https://github.com/dbolya/yolact/issues/222, https://github.com/dbolya/yolact/issues/340.
I have applied Daniel's fix towards this issue with some modifications, and have pushed to the `haotian-dev` branch.

You could try one of the two followings to see if your issue has been resolved:
(1) train using a single GPU
(2) train with the latest code in `haotian-dev` branch with multiple GPUs.

If the issue persists, please **reply directly to #47** (this will be the main thread to deal with related issues for now) with the following information, thanks.
(1) number of GPUs used to train the model
(2) batch size per GPU
(3) learning rate
(4) number of iterations trained
(5) dataset-related information (# of classes, # of training images).

**Comment by ntut108318099** [comment]
Hi, I am training the model by my custom dataset and have pull the latest code on 'haotian-dev' branch.

It might take about 2 days. When it finish,I will evaluate  immediately!!
If it  still have problem,I'll let you know.

BTW,I trained the base model in original config (80 classes , same lr_rate,lr_step ...). 
The result looks great, and it didn't show the error above.

Now,I am training my custom dataset  to watch the error appear or not. 

Here's my custom dataset information:
1. batch_size: 4
2. number of GPUs used to train the model: 1
3. learning rate: 1e-3
4. number of iterations trained: 800000
5. dataset-related information (# of classes, # of training images): 1 classes ,  64050 images .

If I have any new progress, I'll let you know immediately.

Thanks for your hard work.

**Comment by haotian-liu** [comment]
@ntut108318099 Hi thank you for your reply. Yes the original code should work fine when training on a single GPU. The reason you might see a model not trained well might be two reasons:
1. Your batch size is 4, which might not be sufficiently large (default is 8) for training the batchnorm, so it is better to turn it off `cfg.freeze_bn = True`. I might later add this auto freeze_bn into the code base, but you need to do that manually at this point.
2. As your batch size is halved, you should also reduce the learning rate to half.

**Comment by ntut108318099** [comment]
@haotian-liu  Hi, I have finished the model of two versions .
 
I used those models to evaluate,and them still get the same error...
the error still showed : Expected 3 elements in a list but found 2

Did tensorrt transfer need to use specific model or it didn't support one class model ?

**Comment by haotian-liu** [comment]
@ntut108318099 Hi can you provide the lr, batch size you are using? And also the loss figure of tensorboard.

**Comment by ntut108318099** [comment]
@haotian-liu  Here's model information: lr: 5e-4  ,batch_size: 4 . 

About loss figure,I think it's covered by my new test model's log.... 

BTW, I tried to find out which epoch model will cause the error. 
I found the model of 10000 epochs that have already caused the error.    
But when I ran the command: 'disable_tensorrt' , the result seems great.

**Comment by haotian-liu** [comment]
What about the loss figures from tensorboard?

**Comment by haotian-liu** [comment]
Btw how did you create the coco dataset with only person categories? ( if I remembered correctly) and could you share that JSON with me?

**Comment by haotian-liu** [comment]
And also the configs you used (including dataset labels/classes/config you use to train) thanks! I am not sure if it does not support single category conversion using TensorRT ( I think more likely not due to that but cannot confirm as we did not do that experiment)

**Comment by haotian-liu** [comment]
@ntut108318099 For the figure I am not sure if I missed it but I did not see the "new log"?

**Comment by ntut108318099** [comment]
@haotian-liu  I have already sent the email about the file to your email. ( lhtliu@ucdavis.edu ).
Thanks for your patience and assistance ...

**Comment by haotian-liu** [comment]
Hi @ntut108318099, I think I have found the problem (it occurs when there is one detection result, and there will be a dimension mismatch when TensorRT engine is enforced). I have pushed the code to the master branch.

I trained the person config for 20000 iterations with a batch size of `8`, without freezing the _bn_, and setting the learning rate to `5e-4` (this might not be optimal). I converted the model to TensorRT using the configuration reported in the paper (except that I only used 5 images to calibrate). This gives mAP of around 10. This suggests that our model should with single-class instance segmentation task, and I believe training it longer can give us better results (in the paper we train for 800000 iterations with a batch size of 8).

**Comment by ntut108318099** [comment]
Hi,@haotian-liu  OK,I see !! 
I'll pull the newest code to train again. Thanks for your assistance !!

I will take your suggestion to train person_model with a batch_size of 8 , and set the learning rate to 5e-4. It will train about 2 days. Then , I will evaluate model . If everything is okay , I'll let you know !!

**Comment by haotian-liu** [comment]
@ntut108318099 Hi that sounds good. The LR to be 5e-4 might not be optimal so if you want the best performance, you might need to search the hyper-parameters.
